### PR TITLE
Add Feature Copy Buttons

### PR DESCRIPTION
**Pull request by clavinjune**
## Motivation

Show copy buttons on new-webapp as requested on #665 

## Modification

- Add `handleCopyApiUrl`, `handleCopyWebUrl`, `handleCopyAsCliCommand`, `handleCopyAsCurlCommand` function in FileList
- Show toast on success / error
- Add `resetState` action on messageSlice
- Add tests

## Result

image

**On Success**

image

**On Error**

image

### CODE REVIEW

**Review by ikhoon — COMMENTED** [review]
Thanks for your interest in our project. 🙇‍♂️

File: webapp/src/dogma/features/file/FileList.tsx [hunk 1]
```diff
@@ -49,6 +55,18 @@ const FileList = <Data extends object>({ data, projectName, repoName }: FileList
               Edit
             </Button>
           </WrapItem>
+          <WrapItem>
+            <Button
+              leftIcon={<CopyIcon />}
+              colorScheme="gray"
+              size="sm"
+              onClick={() =>
+                handleCopyApiUrl(`/api/v1/projects/${projectName}/repos/${repoName}/contents${info.getValue()}`)
```
[the remark this answers is not in the document]
Comment (clavinjune) [reply webapp/src/dogma/features/file/FileList.tsx]: on it @ikhoon

[the remark this answers is not in the document]
Comment (clavinjune) [reply webapp/src/dogma/features/file/FileList.tsx]: added @ikhoon.
Also I wanna ask, is it better to have this kind of UI instead?

image

File: webapp/src/dogma/features/file/FileList.tsx [hunk 2]
```diff
@@ -11,9 +11,17 @@ export type FileListProps<Data extends object> = {
   data: Data[];
   projectName: string;
   repoName: string;
+  handleCopyApiUrl: Function;
+  handleCopyAsCurlCommand: Function;
```
Comment (rainy789) [inline webapp/src/dogma/features/file/FileList.tsx:15]: Let's add some tests for these. 😄 
Here's an example for `handleCopyApiUrl`.
1) Modify the following in `FileList.test.tsx`.
```
    expectedProps = {
      data: mockfileList,
      projectName: 'ProjectAlpha',
      repoName: 'repo1',
      handleCopyApiUrl: jest.fn(),
    };
```
2) Add the unit test
```
  it('calls handleCopyApiUrl when copy API URL button is clicked', () => {
    const { getAllByText } = render(<FileList {...expectedProps} />);
    const firstButton = getAllByText('Copy API URL', { selector: 'button' })[0];
    fireEvent.click(firstButton);
    expect(expectedProps.handleCopyApiUrl).toHaveBeenCalledTimes(1);
  });
```
3) Don't forget to import 
```import { fireEvent, render } from '@testing-library/react';```

4) Similarly for `handleCopyAsCurlCommand`

Comment (clavinjune) [reply webapp/src/dogma/features/file/FileList.tsx:15]: on it @rainy789

Comment (clavinjune) [reply webapp/src/dogma/features/file/FileList.tsx:15]: added ya @rainy789 

image

File: webapp/src/dogma/features/file/FileList.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply webapp/src/dogma/features/file/FileList.tsx]: >  Also I wanna ask, is it better to have this kind of UI instead?

It looks better! In addition to those URLs, copying a web URL for the file would be useful.
- API URL
- Web URL
- cURL

[the remark this answers is not in the document]
Comment (ikhoon) [reply webapp/src/dogma/features/file/FileList.tsx]: As a similar work, Chrome network monitoring console provides a menu for copy.
image

[the remark this answers is not in the document]
Comment (ikhoon) [reply webapp/src/dogma/features/file/FileList.tsx]: I don't think we need to support fetch or other things at the moment. We may add a code snippet for our go cli client. https://line.github.io/centraldogma/client-cli.html

File: webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx [hunk 3]
```diff
@@ -47,9 +50,37 @@ const RepositoryDetailPage = () => {
...
+  const constructApiUrl = (project: string, repo: string, path: string): string => {
+    return `${window.location.origin}/api/v1/projects/${project}/repos/${repo}/contents${path}`;
+  };
+  const handleCopyApiUrl = async (project: string, repo: string, path: string) => {
+    const apiUrl: string = constructApiUrl(project, repo, path);
+    copyToClipboard(apiUrl);
+  };
+  const handleCopyAsCurlCommand = async (project: string, repo: string, path: string) => {
+    const apiUrl: string = constructApiUrl(project, repo, path);
+    const curlCommand = `curl -XGET "${apiUrl}"`;
```
Comment (ikhoon) [inline webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx:80]: We need to set `sessionId` as an access token to Authorization header to access resources using REST API.
```
curl -H "Authorization: Bearer access-token"
```

Comment (clavinjune) [reply webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx:80]: on it

File: webapp/src/dogma/features/file/FileList.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (clavinjune) [reply webapp/src/dogma/features/file/FileList.tsx]: on it

File: webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx (hunk 3 above)
Comment (clavinjune) [reply webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx]: added @ikhoon

File: webapp/src/dogma/features/file/FileList.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (clavinjune) [reply webapp/src/dogma/features/file/FileList.tsx]: added @ikhoon

[the remark this answers is not in the document]
Comment (clavinjune) [reply webapp/src/dogma/features/file/FileList.tsx]: image

File: webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx (hunk 3 above)
Comment (clavinjune) [reply webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx]: This is the copy result:

```shell
curl -XGET "http://localhost:3000/api/v1/projects/abcd/repos/repo1/contents/0-dwarf-hotot" \
-H "Authorization: Bearer <access-token>"
```

File: webapp/src/dogma/features/file/FileList.tsx [hunk 4]
```diff
@@ -11,9 +11,21 @@ export type FileListProps<Data extends object> = {
...
 
-const FileList = <Data extends object>({ data, projectName, repoName }: FileListProps<Data>) => {
+const FileList = <Data extends object>({
+  data,
+  projectName,
+  repoName,
+  handleCopyApiUrl,
+  handleCopyWebUrl,
+  handleCopyAsCliCommand,
+  handleCopyAsCurlCommand,
```
Comment (ikhoon) [inline webapp/src/dogma/features/file/FileList.tsx:27]: How about providing this function as an object? `CopySupport`?
And use it like:
```
copySupport.handleApiUrl(...)

```

File: webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx [hunk 5]
```diff
@@ -47,9 +50,52 @@ const RepositoryDetailPage = () => {
...
+    } catch (err) {
+      const error: string = ErrorHandler.handle(err);
+      await dispatch(createMessage({ title: 'failed to copy to clipboard', text: error, type: 'error' }));
+    } finally {
+      dispatch(resetState());
+    }
+  };
+
+  const constructApiUrl = (project: string, repo: string, path: string): string => {
+    return `${window.location.origin}/api/v1/projects/${project}/repos/${repo}/contents${path}`;
```
Comment (ikhoon) [inline webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx:72]: It is necessary to provide `revision` as a parameter of the API.
```suggestion
    let apiUrl = `${window.location.origin}/api/v1/projects/${project}/repos/${repo}/contents${path}`;
    if (revision !== 'head') {
      apiUrl += `?revision=${revision}`;
    }
```

File: webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx [hunk 6]
```diff
@@ -47,9 +50,52 @@ const RepositoryDetailPage = () => {
...
+  };
+
+  const handleCopyApiUrl = async (project: string, repo: string, path: string) => {
+    const apiUrl: string = constructApiUrl(project, repo, path);
+    copyToClipboard(apiUrl);
+  };
+
+  const handleCopyWebUrl = async (project: string, repo: string, path: string) => {
+    const webUrl = `${window.location.origin}/app/projects/${project}/repos/${repo}/files/head${path}`;
```
Comment (ikhoon) [inline webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx:81]: Should it be?
```suggestion
    const webUrl = `${window.location.origin}/app/projects/${project}/repos/${repo}/files/${revision}${path}`;
```

File: webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx [hunk 7]
```diff
@@ -47,9 +50,52 @@ const RepositoryDetailPage = () => {
...
+    copyToClipboard(webUrl);
+  };
+
+  const handleCopyAsCliCommand = async (project: string, repo: string, path: string) => {
+    const cliCommand = `dogma --connect "${window.location.origin}" \\
+--token "<access-token>" \\
+cat ${project}/${repo}${path}`;
```
Comment (ikhoon) [inline webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx:88]: Should we add `--revision ${revision}` option when the revision is not `head`?

**Review by ikhoon — COMMENTED** [review]
The UI looks good. Left comments about revision in the URL

File: webapp/src/dogma/features/file/FileList.tsx [hunk 8]
```diff
@@ -49,6 +61,27 @@ const FileList = <Data extends object>({ data, projectName, repoName }: FileList
...
+                  API URL
+                </MenuItem>
+                <MenuItem onClick={() => handleCopyWebUrl(projectName, repoName, info.getValue())}>
+                  Web URL
+                </MenuItem>
+                <MenuItem onClick={() => handleCopyAsCliCommand(projectName, repoName, info.getValue())}>
+                  CLI Command
+                </MenuItem>
+                <MenuItem onClick={() => handleCopyAsCurlCommand(projectName, repoName, info.getValue())}>
+                  cURL Command
```
Comment (ikhoon) [inline webapp/src/dogma/features/file/FileList.tsx:80]:
```suggestion
                  CLI command
                </MenuItem>
                <MenuItem onClick={() => handleCopyAsCurlCommand(projectName, repoName, info.getValue())}>
                  cURL command
```

File: webapp/src/dogma/features/file/FileList.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply webapp/src/dogma/features/file/FileList.tsx]: It looks really nice!

File: webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx [hunk 9]
```diff
@@ -47,9 +51,64 @@ const RepositoryDetailPage = () => {
...
+      copyToClipboard(webUrl);
+    },
+
+    async handleAsCliCommand(project: string, repo: string, path: string) {
+      let cliCommand = `dogma --connect "${window.location.origin}" \\
+--token "<access-token>" \\
+cat ${project}/${repo}${path}`;
+
+      if (revision !== 'head') {
+        cliCommand += ` --revision=${revision}`;
```
Comment (ikhoon) [inline webapp/src/pages/app/projects/[projectName]/repos/[repoName]/list/[revision].tsx:98]: CLI uses a space as a separator.
```suggestion
        cliCommand += ` --revision ${revision}`;
```

**Review by ikhoon — APPROVED** [review]
Thanks a lot! 💯🚀
I'm looking forward to your next step. 😄

**Pull request merged**